src/Language/Pirouette/QuasiQuoter/Syntax.hs

```diff
@@ -182,56 +275,80 @@ parseType = label "Type" $ makeExprParser pAtom [[InfixL pApp], [InfixR pFun]]
     pFun = TyFun <$ symbol "->"
 
     pAll :: Parser (Ty lang)
-    pAll = label "pAll" $ do
-      try (symbol "all")
-      parseBinder TyAll (parseBinderNames (ident @lang) (parseTyOf >> parseKind)) (symbol "." >> parseType)
+    pAll = P.label "pAll" $ do
+      P.try (symbol "forall")
+      parseBinder
+        TyAll
+        (some (parseTypeVarKind @lang))
+        (symbol "." >> parseType)
 
     pLam :: Parser (Ty lang)
-    pLam = label "pLam" $ do
-      try (symbol "\\")
-      parseBinder TyLam (parseBinderNames (ident @lang) (parseTyOf >> parseKind)) (symbol "." >> parseType)
+    pLam = P.label "pLam" $ do
+      P.try (symbol "\\")
+      parseBinder
+        TyLam
+        (some (parseTypeVarKind @lang))
+        (symbol "." >> parseType)
 
     pAtom =
-      try $
+      P.try $
         asum
           [ pLam,
             pAll,
             parseTypeAtom
           ]
 
+-- | Parse type atoms among type variables, builtins, and types declared in the
+-- environment
 parseTypeAtom :: forall lang. (LanguageParser lang) => Parser (Ty lang)
 parseTypeAtom =
-  label "Type atom" $
+  P.label "Type atom" $
     asum
-      [ TyVar <$> try (ident @lang),
-        TyBase <$> try (parseBuiltinType @lang),
-        TyFree <$> try (typeName @lang),
+      [ TyVar <$> P.try (ident @lang),
+        TyBase <$> P.try (parseBuiltinType @lang),
+        TyFree <$> P.try (typeName @lang),
         parens parseType
       ]
 
+-- TODO Keep in this module?
 exprBinApp :: BuiltinTerms lang -> Expr lang -> Expr lang -> Expr lang
 exprBinApp f x = ExprApp (ExprApp (ExprBase f) x)
 
+-- | Parse an expression. The following features are supported:
+--
+-- - Type abstraction: `/\ (a : * -> *) (b : *) . `
+-- - Lambdas: `\ (x : Maybe a) (y : a) . `
+-- - If/then/else: `if @resultType (x == y) then expr1 else expr2`
+-- - Case statements:
+--   `case @type @resultType x of {pattern1 -> expr1 ; pattern2 -> expr2}`
+-- - Bottom: `bottom @resultType`
+-- - Type and term application: `f @a @b x @c y z`
```

Comment:
Code blocks in haddock are written using `@`: https://haskell-haddock.readthedocs.io/en/latest/markup.html#code-blocks

```suggestion
-- - Type abstraction: @/\ (a : * -> *) (b : *) .@
-- - Lambdas: @\ (x : Maybe a) (y : a) . @
-- - If/then/else: @if \@resultType (x == y) then expr1 else expr2@
-- - Case statements:
--   @case \@type \@resultType x of {pattern1 -> expr1 ; pattern2 -> expr2}@
-- - Bottom: @bottom \@resultType@
-- - Type and term application: @f \@a \@b x \@c y z@
```